subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c

```diff
@@ -14,90 +14,103 @@
- *        Radio peripheral. It is used in macros only e.g. UTIL_LISTIFY that
- *        evaluate indices in an inclusive way.
+/* Devicetree node identifier for the radio node. */
+#define RADIO_NODE DT_NODELABEL(radio)
+
+/* Value to set for unconnected antenna GPIO pins. */
+#define DFE_PSEL_NOT_SET 0xFF
+/* Number of PSEL_DFEGPIO[n] registers in the radio peripheral. */
+#define MAX_DFE_GPIO 8
+/* Run a macro 'fn' on each available DFE GPIO index, from 0 to
+ * MAX_DFE_GPIO-1, with the given parenthesized separator.
  */
-#define DF_PSEL_GPIO_NUM_MAX_IDX 7
+#define FOR_EACH_DFE_GPIO(fn, sep) \
+	FOR_EACH(fn, sep, 0, 1, 2, 3, 4, 5, 6, 7)
 
-/* @brief Direction Finding antenna matrix configuration */
+/* Direction Finding antenna matrix configuration */
 struct df_ant_cfg {
 	uint8_t ant_num;
 	/* Selection of GPIOs to be used to switch antennas by Radio */
-	uint8_t dfe_gpio[DF_PSEL_GPIO_NUM];
+	uint8_t dfe_gpio[MAX_DFE_GPIO];
 };
 
-#define RADIO DT_NODELABEL(radio)
-#define DFE_GPIO_PSEL(idx)                                                \
-	COND_CODE_1(DT_NODE_HAS_PROP(RADIO, dfegpio##idx##_gpios),        \
-		   (NRF_DT_GPIOS_TO_PSEL(RADIO, dfegpio##idx##_gpios)),   \
-		   (DF_GPIO_PIN_NOT_SET))
+#define DFE_GPIO_PSEL(idx)					  \
+	NRF_DT_GPIOS_TO_PSEL_OR(RADIO_NODE, dfegpio##idx##_gpios, \
+				DFE_PSEL_NOT_SET)
 
 #define DFE_GPIO_PIN_DISCONNECT (RADIO_PSEL_DFEGPIO_CONNECT_Disconnected << \
 				 RADIO_PSEL_DFEGPIO_CONNECT_Pos)
 
-#define COUNT_GPIO(idx, _) + DT_NODE_HAS_PROP(RADIO, dfegpio##idx##_gpios)
-#define DFE_GPIO_NUM        (UTIL_LISTIFY(DF_PSEL_GPIO_NUM, COUNT_GPIO, _))
+#define HAS_DFE_GPIO(idx) DT_NODE_HAS_PROP(RADIO_NODE, dfegpio##idx##_gpios)
 
-/* DFE_GPIO_NUM_IS_ZERO is required to correctly compile COND_CODE_1 in
- * DFE_GPIO_ALLOWED_ANT_NUM macro. DFE_GPIO_NUM does not expand to literal 1
- * So it is not possible to use it as a conditional in COND_CODE_1 argument.
+/* The number of dfegpio[n]-gpios properties which are set. */
+#define DFE_GPIO_NUM (FOR_EACH_DFE_GPIO(HAS_DFE_GPIO, (+)))
+
+/* The minimum number of antennas required to enable antenna switching. */
+#define MIN_ANTENNA_NUM 2
+
+/* The maximum number of antennas supported by the number of
+ * dfegpio[n]-gpios properties which are set.
  */
 #if (DFE_GPIO_NUM > 0)
-#define DFE_GPIO_NUM_IS_ZERO 1
+#define MAX_ANTENNA_NUM BIT(DFE_GPIO_NUM)
 #else
-#define DFE_GPIO_NUM_IS_ZERO EMPTY
+#define MAX_ANTENNA_NUM 0
 #endif
 
-#define DFE_GPIO_ALLOWED_ANT_NUM COND_CODE_1(DFE_GPIO_NUM_IS_ZERO,       \
-					     (BIT(DFE_GPIO_NUM)), (0))
+/*
+ * Check that we have an antenna switch pattern.
+ */
+
+#define HAS_PDU_ANTENNA DT_NODE_HAS_PROP(RADIO_NODE, dfe_pdu_antenna)
+
+#define PDU_ANTENNA DT_PROP_OR(RADIO_NODE, dfe_pdu_antenna, 0)
+
+BUILD_ASSERT(HAS_PDU_ANTENNA,
+	     "Missing antenna pattern used to select antenna for PDU Tx. "
+	     "Set the dfe-pdu-antenna devicetree property.");
```

Comment:
Move it back into 
```
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
```
please. 

Also make `radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()` functions conditionally compilable. It will decrease default compilation code coverage but it will give possibility to build DF beacon application working in AoA mode. The APP will not need DTS overlay for DFE configuration.
If `BUILD_ASSERT` is not conditional code, even simple app, transmitting CTE in AoA mode (that does not use antenna switching feature of DFE), requires an DTS overlay.

Reply:
> Also make `radio_df_pdu_antenna_switch_pattern_get()` and `radio_df_ant_switch_pattern_set()` functions conditionally compilable.

OK, I will put them under the same ifdef as the BUILD_ASSERT.

However, usage of these functions is different and I'm not sure that is always correct.

- radio_df_pdu_antenna_switch_pattern_set() is called from lll_df.c, and only if CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y or CONFIG_BT_CTLR_DF_ANT_SWITCH_RX=y
- radio_df_pdu_antenna_switch_pattern_get() is only called from lll_sync.c, but only if CONFIG_BT_CTLR_DF_SCAN_CTE_RX=y

As far as I can tell, CONFIG_BT_CTLR_DF_SCAN_CTE_RX doesn't depend on either CONFIG_BT_CTLR_DF_ANT_SWITCH_TX or CONFIG_BT_CTLR_DF_ANT_SWITCH_RX. So I'm not sure lll_sync.c is doing the right thing there.

Is this actually OK?

Reply:
>>Also make radio_df_pdu_antenna_switch_pattern_get() and radio_df_ant_switch_pattern_set() functions conditionally compilable.

> I'm not sure lll_sync.c is doing the right thing there.

Yep, getting a link error in CI:

```
/workdir/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:400: undefined reference to `radio_df_pdu_antenna_switch_pattern_get'
```

I'm not sure what changes you'd like to see here in terms of making radio_df_pdu_antenna_switch_pattern_get() conditionally compiled.

Reply:
Ah... I'm sorry. In such situation move the the radio_df_pdu_antenna_switch_pattern_get function out of `#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)` conditional compilation block and change like this:
```
uint8_t radio_df_pdu_antenna_switch_pattern_get(void)
{
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
	return PDU_ANTENNA;
#else
	return 0;
#endif
}
```

Reply:
OK, understood. I'd like to suggest a slightly different approach: moving both of these lines outside of the conditional compilation block:

```
#define PDU_ANTENNA DT_PROP_OR(RADIO_NODE, dfe_pdu_antenna, 0)

uint8_t radio_df_pdu_antenna_switch_pattern_get(void)
{
	return PDU_ANTENNA;
}
```

I think that's a little easier to read. LMK if you disagree.

Reply:
Yes, it looks better. I'll remember to use `DT_PROP_OR` in future for such cases.